core/FusionDecode.cpp

```diff
@@ -0,0 +1,220 @@
+// TODO: add Condor header - this is a new file
+// contact jeff at condor
+#include "FusionTypes.hpp"
+#include "Decode.hpp"
+
+#include "sparta/events/StartupEvent.hpp"
+#include "sparta/utils/LogUtils.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <set>
+#include <map>
+#include <vector>
+
+using namespace std;
+using namespace fusion;
+
+namespace olympia
+{
+// -------------------------------------------------------------------
+// Remove the ghost fusion ops
+// Kept, but currently not called
+// -------------------------------------------------------------------
+void Decode::whereIsEgon(InstGroupPtr &insts,uint32_t &numGhosts)
+{
+    if(insts == nullptr) {
+        //DLOG("whereIsEgon() called with insts null");
+        return;
+    }
+
+    //DLOG("BEFORE # --------------------------------------");
+    //infoInsts(cout,insts);
+
+    for(auto it = insts->begin(); it != insts->end(); ) {
+        if ((*it)->getExtendedStatus()
+                   == olympia::Inst::Status::FUSION_GHOST)
+        {
+            ++numGhosts;
+            it = insts->erase(it);
+        } else {
+            ++it;
+        }
+    }
+    //DLOG("AFTER  # --------------------------------------");
+    //infoInsts(cout,insts);
+    //DLOG("DONE   # --------------------------------------");
+}
+// -------------------------------------------------------------------
+// -------------------------------------------------------------------
+void Decode::processMatches(MatchInfoListType &matches,
+                            InstGroupPtr &insts,
+                            const InstUidListType &inputUids)
+{
+    if(matches.size() == 0) return;
+
+    //Kept for upcoming additional work on reducing ROB entry requirements
+    //DLOG("PROCESS MATCHES "<<matches.size());
+    //DLOG("FUSION: REF   UIDS:");
+    //fuser->infoUids(cout,inputUids,"before "); cout<<endl;
+    //fuser->info(cout,matches,inputUids); cout<<endl;
+    //size_t idx = 0;
+
+    for(const auto &m : matches)
+    {
+        //Kept for upcoming additional work on reducing ROB entry requirements
+        //DLOG("FUSION: M("<<idx++<<")  UIDS:");
+        //string gap(m.startIdx*5,' ');
+        //DLOG(gap);
+        //for(const auto &mu:m.matchedUids) DLOG(" 0x"<<hex<<setw(2)<<mu);
+        //DLOG("\n");
+
+        InstGroup::iterator q = insts->begin();
+        std::advance(q,m.startIdx);
+        if((*q)->getExtendedStatus() == Inst::Status::UNMOD)
+        {
+
+            (*q)->setExtendedStatus(Inst::Status::FUSED);
+            ++fusion_num_fuse_instructions_;
+            //Kept for upcoming additional work on reducing ROB 
+            //entry requirements. This manages the sequential expectations
+            //of the ROB for program ID.
+            //(*q)->setProgramIDIncrement(m.size());
+
+            for(size_t i=1;i<m.size();++i) {
+                std::advance(q,1);
```

Comment:
`++q`?  :) 

You probably should ensure that `q != insts.end()` unless that's not possible

Reply:
advance was a carry over from before it was known to always be a constant increment.

fixed

it is not possible to exceed the range but it is not obvious from this function,  so I added an assert.